core/src/main/scala/kafka/network/SocketServer.scala

```diff
@@ -486,12 +487,14 @@ private[kafka] class Processor(val id: Int,
   private def processCompletedReceives() {
     selector.completedReceives.asScala.foreach { receive =>
       try {
-        val channel = selector.channel(receive.source)
+        val openChannel = selector.channel(receive.source)
+        val channel = if (openChannel != null) openChannel else selector.closedChannel(receive.source)
```

Comment:
Oh, I just realised that we also need the principal from the channel.

By the way, we had the following issue reported a while back: https://issues.apache.org/jira/browse/KAFKA-3986, but I could not figure out under which cases we would end up with a `null` channel here. If you have, I'd be interested to know. :)